frontend/app/ui_components/expandable-search.js

```diff
@@ -0,0 +1,73 @@
+//-- copyright
+// OpenProject is a project management system.
+// Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License version 3.
+//
+// OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+// Copyright (C) 2006-2013 Jean-Philippe Lang
+// Copyright (C) 2010-2013 the ChiliProject Team
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+//
+// See doc/COPYRIGHT.rdoc for more details.
+//++
+
+module.exports = function(ENTER_KEY) {
+  return {
+    restrict: 'E',
+    replace: true,
+    link: function(scope, elem) {
+      var btn = elem.find('a'),
+        input = elem.find('input'),
+        setCollapsed;
+
+      // Search is collapsed initially
+      scope.collapsed = true;
+      setCollapsed = function(collapsedState) {
+          scope.collapsed = collapsedState;
+          scope.$apply();
+        };
+
+      btn.on('click mousedown focus keypress', function(evt) {
+        if (scope.collapsed === true) {
+          setCollapsed(false);
+          // Avoid access key focus
+          setTimeout(function() { input.focus(); });
+
+          // Hide on lost focus
+          elem.on('focusout', function() {
+            // Allow DOM to propagate new focus
+            setTimeout(function() {
+              // Hide unless icon clicked
+              if (elem.find(':active,:focus').length === 0) {
+                setCollapsed(true);
+                $(elem).off('focusout');
+              }
+            }, 10);
+          });
+          evt.preventDefault();
+        } else {
+          // Submit only when clicked or enter on btn
+          if (evt.type === 'mousedown' ||
+             (evt.type === 'keypress' && evt.which === ENTER_KEY)) {
+            elem.closest('form').submit();
```

Comment:
I can't help you on that one either. Maybe @myabc has any valuable input?

I can think of two ways out:

1. the one you have chosen: The form is not part of the directive, but you expect the directive to be used within a form (still don't know if `elem.closest` is the best possible solution here)
2. What I could also imagine is to include the form **in** your directive. After all you need that form, so it might be a good idea to make it a part of your component. This should ease finding it, but might introduce other problems I did not yet see.

I would merge it either way, but I am also happy to receive further feedback beforehand.

@oliverguenther Tell me when you feel ready enough ;-)